sdk/src/main/scala/io/horizen/certificatesubmitter/AbstractCertificateSubmitter.scala

```diff
@@ -269,33 +270,44 @@ abstract class AbstractCertificateSubmitter[
       val wallet = sidechainNodeView.vault
-      val privateKeysWithIndexes = signersPublicKeys.map(signerPublicKey => wallet.secret(signerPublicKey)).zipWithIndex.filter(_._1.isDefined).map {
+      signersPublicKeys.map(signerPublicKey => wallet.secret(signerPublicKey)).zipWithIndex.filter(_._1.isDefined).map {
         case (secretOpt, idx) => (secretOpt.get.asInstanceOf[SchnorrSecret], idx)
       }
+   }
 
-      val remainingKeys = signersPublicKeys.zipWithIndex.filterNot(key_index => privateKeysWithIndexes.map(_._2).contains(key_index._2))
-      (signaturesFromEnclave(messageToSign, remainingKeys)
-        ++ privateKeysWithIndexes.map {
-        case (secret, pubKeyIndex) => CertificateSignatureInfo(pubKeyIndex, secret.sign(messageToSign))
-      })
-    }
+    val privateKeysWithIndexes = Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(getSignersLocalPrivateKeys), timeoutDuration)
+      .asInstanceOf[Seq[(SchnorrSecret, Int)]]
+    val remainingKeys = signersPublicKeys.zipWithIndex.filterNot(key_index => privateKeysWithIndexes.map(_._2).contains(key_index._2))
+    (signaturesFromEnclave(messageToSign, remainingKeys)
+      ++ privateKeysWithIndexes.map {
+      case (secret, pubKeyIndex) => CertificateSignatureInfo(pubKeyIndex, secret.sign(messageToSign))
+    })
 
-    Await.result(sidechainNodeViewHolderRef ? GetDataFromCurrentView(getSignersPrivateKeys), timeoutDuration)
-      .asInstanceOf[Seq[CertificateSignatureInfo]]
   }
 
   def signaturesFromEnclave(messageToSign: Array[Byte], indexedPublicKeys: Seq[(SchnorrProposition, Int)]): Seq[CertificateSignatureInfo] = {
     if (!secureEnclaveApiClient.isEnabled) return Seq()
-    val signaturesFromEnclaveFuture = secureEnclaveApiClient.listPublicKeys()
-      .map(managedKeys => indexedPublicKeys.filter(key_index => managedKeys.contains(key_index._1)))
-      .map(_.map(secureEnclaveApiClient.signWithEnclave(messageToSign, _)))
-      .map(Future.sequence(_))
-      .flatten
-
-    Try(Await.result(signaturesFromEnclaveFuture, timeoutDuration).flatten)
-      .getOrElse(Seq())
+    val keysFromEnclaveFuture = secureEnclaveApiClient.listPublicKeys()
+    val listOfKeys = Try(Await.result(keysFromEnclaveFuture, secureEnclaveRequestTimeout))
+      .map(managedKeys => indexedPublicKeys.filter(key_index => managedKeys.contains(key_index._1))).getOrElse(Seq())
+
+    val steps = listOfKeys.grouped(numOfParallelRequests)
+    val listOfSignatures = for (step <- steps) yield {
+      val listOfSignaturesFutures = step.map(secureEnclaveApiClient.signWithEnclave(messageToSign, _))
+
+      Try(Await.result(Future.sequence(listOfSignaturesFutures), secureEnclaveRequestTimeout).flatten)
+        .getOrElse {
+          val signaturesOpt = for {
+            sigFutureResult <- listOfSignaturesFutures
+            if sigFutureResult.isCompleted && sigFutureResult.value.get.isSuccess
+          } yield sigFutureResult.value.get.get
```

Comment:
if you want, you can do this to avoid if checks and .value.get.get
```
  val signaturesOpt =
    listOfSignaturesFutures
      .map(_.value) // turn Future into Option[Try[Option[CertificateSignatureInfo]]]
      .collect {
        case Some(Success(sigResult)) => sigResult
      }
```